src/query/pipeline/core/src/finished_chain.rs

```diff
@@ -0,0 +1,388 @@
+        false
+    }
+
+    fn apply(self: Box<Self>, info: &ExecutionInfo) -> Result<()>;
+}
+
+struct ApplyState {
+    is_always: bool,
+    successfully: bool,
+    elapsed: Duration,
+    location: &'static Location<'static>,
+}
+
+pub struct FinishedCallbackChain {
+    chain: VecDeque<(&'static Location<'static>, Box<dyn Callback>)>,
+}
+
+impl FinishedCallbackChain {
+    pub fn create() -> FinishedCallbackChain {
+        FinishedCallbackChain {
+            chain: VecDeque::new(),
+        }
+    }
+
+    pub fn push_front(&mut self, location: &'static Location<'static>, f: Box<dyn Callback>) {
+        self.chain.push_front((location, f))
+    }
+
+    pub fn push_back(&mut self, location: &'static Location<'static>, f: Box<dyn Callback>) {
+        self.chain.push_back((location, f))
+    }
+
+    pub fn apply(&mut self, mut info: ExecutionInfo) -> Result<()> {
+        let chain = std::mem::take(&mut self.chain);
+
+        let mut states = Vec::with_capacity(chain.len());
+        let mut callbacks = Vec::with_capacity(chain.len());
+        let mut always_callbacks = Vec::with_capacity(chain.len());
+
+        for (location, callback) in chain.into_iter() {
+            if !callback.always_call() {
+                callbacks.push((location, callback));
+            } else {
+                always_callbacks.push((location, callback));
+            }
+        }
+
+        for (location, callback) in callbacks {
+            let instant = Instant::now();
+            if let Err(cause) = callback.apply(&info) {
+                states.push(ApplyState {
+                    location,
+                    is_always: false,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                });
+
+                info.res = Err(cause.clone());
+                Self::apply_always(info, states, always_callbacks);
+
+                return Err(cause);
+            }
+
+            states.push(ApplyState {
+                location,
+                is_always: false,
+                successfully: true,
+                elapsed: instant.elapsed(),
+            });
+        }
+
+        Self::apply_always(info, states, always_callbacks);
+        Ok(())
+    }
+
+    fn apply_always(
+        info: ExecutionInfo,
+        mut states: Vec<ApplyState>,
+        always_callbacks: Vec<(&'static Location<'static>, Box<dyn Callback>)>,
+    ) {
+        for (location, always_callback) in always_callbacks {
+            let instant = Instant::now();
+            states.push(match always_callback.apply(&info) {
+                Ok(_) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: true,
+                    elapsed: instant.elapsed(),
+                },
+                Err(_cause) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: false,
```

Comment:
maybe not useful for now, but the cause can be recorded?

Reply:
The cause may have been recorded in the log.